feat: implement auto approval with pre-receive hooks

fabiovincenzi opened the pull request:
Summary

This PR is based on #11 and it introduces automatic approval and rejection for push actions by using pre-receive hooks.

Changes

1. Backend

   - Introduced `autoApproved` and `autoRejected` properties in `Action.js` to track automatic approvals and rejections.
   - Added `setAllowAutoApproval()` and `setAutoRejection()` methods to mark actions as auto-approved or auto-rejected.
   - Implemented `attemptAutoApproval()` and `attemptAutoRejection()` to automatically authorize or reject actions when applicable.
   - Updated pre-receive hook processing to recognize when a push should be auto-approved or auto-rejected.

2. Frontend

   - Updated `PushDetails.jsx`:
     - Disabled interaction with approval icons for auto-approved actions.
     - Adjusted cursor and opacity to indicate non-interactive elements.
     - Modified the attestation modal behavior to prevent opening for auto-approved action.
 
<img width="239" alt="Screenshot 2025-03-25 at 16 37 10" src="https://github.com/user-attachments/assets/faac613d-7af5-424e-bb68-dfadbf32255c" />


3. Tests & Pre-Receive Hook Changes

   - Added tests for automatic approvals and rejections in `chain.test.js` and `preReceive.test.js`.
   - Updated pre-receive hooks to return appropriate status codes for auto-approved, manually approved, and rejected pushes.
   - Verified that database operations (`authorise` and `reject`) are correctly called for auto-approved and auto-rejected actions.

4. Documentation
   - Created `pre-receive.md` documentation to clarify the functionality of the pre-receive hook.

JamieSlome commented:
@fabiovincenzi - can we take a look at the code coverage warning above and pull up (💪) to 100%?

JamieSlome commented on docs/pre-receive.md:
```diff
@@ -0,0 +1,52 @@
+
+## Overview
+
+The `pre-receive` hook is a critical component of the Git Proxy system. It is executed before changes are accepted into a repository. This hook allows for custom logic to validate or reject incoming changes based on specific criteria, ensuring that only valid and authorized changes are pushed to the repository.
```
```suggestion
The `pre-receive` hook is a critical component of the GitProxy system. It is executed before changes are accepted into a repository. This hook allows for custom logic to validate or reject incoming changes based on specific criteria, ensuring that only valid and authorized changes are pushed to the repository.
```

JamieSlome commented on docs/pre-receive.md:
Can we find a suitable home for this in the `website/docs` folder? This corresponds to [git-proxy.finos.org](https://git-proxy.finos.org) 👍

JamieSlome commented on src/ui/views/PushDetails/PushDetails.jsx:
Any chance we can get a Cypress test for this?

JamieSlome reviewed (commented):
Great PR 👍 Just a couple of requested changes. Only general question from me is whether this is backwards compatible, i.e. does historical data generated by GitProxy require this flag or is purely forward facing?

fabiovincenzi commented:
> Great PR 👍 Just a couple of requested changes. Only general question from me is whether this is backwards compatible, i.e. does historical data generated by GitProxy require this flag or is purely forward facing?

Thank you, @JamieSlome! I just started addressing comments, I would say it is backward compatible, all the old pushes wouldn't have any autoApproved or autoRejected flag and they are going to be treated normally as always

fabiovincenzi commented:
@JamieSlome: It's all set up. Also, increased code coverage. I think for Action.js is not necessary, but I am happy to discuss if you think we need something

JamieSlome reviewed (approved):
LGTM! 🍰
